helix-term/src/commands.rs

```diff
@@ -4169,16 +4169,23 @@ pub fn completion(cx: &mut Context) {
 
     let (view, doc) = current!(cx.editor);
 
+    let savepoint = if let Some(CompleteAction::Selected { savepoint }) = &cx.editor.last_completion
+    {
+        savepoint.clone()
+    } else {
+        doc.savepoint(view)
+    };
+
```

Comment:
Can you explain what this does exactly (maybe add a comment then)? I still have not really understood how this'll happen (`last_completion` is `Some`), why it's necessary, and if it's even correct (I think when an item is selected and the completion is requested again, this should be included/applied I guess, but this seems to happen somewhere else already as discovered below).

I tried to trigger this case by pressing `C-x` while already a completion menu is open with a selected item, to trigger this case, but `last_completion` was already `None`, as the selected completion is applied when the completion menu doesn't swallow (that behavior seems to be added in this PR as well below in `editor.rs`).

Reply:
There are basically two versions of the document:
* the version stored in the server
* the version stored in the client

In theory, these are synced but they can become out of sync. That's for example why we have to check version numbers before applying workspace edits. This PR introduced the concept of virtual edits which are not sent to the server. These are used when selecting a completion. This purposefully desynchronizes the two document versions so that the selected completion doesn't affect the server-side completion filtering (which can produce weird results or missing completion items). 

The savepoint in CompleteAction::Selected { savepoint } represents the last state send to the server.  Before we interact with the LSP in any way we must always reset to this state. This is not a big problem for async requests/notifiactions from the server as we already check the version number anyway. It's also not a problem for commands trigger by keypresses. Because every keypress will automatically apply the completion (which synchronizes the server). This is also why `c-x` doesn't trigger this branch as you are pressing a key.

However, what is a problem is places where we send requests to the server asynchronously so not directly triggered by a keypress. Completion requests are sent automatically on idle-timeout. The idle timeout does not cause the completion to be automatically accepted tough and can therefore cause invalid positions to be sent to the server. With a robust server this usually just means no/wrong completions but the problem is then that the completions it sends back will (as always) refer to the state of the document when the completion request was sent. However, as the server side state is (purposefully) out of sync that means that the server will send back potentially invalid edits which would cause similar problems as fixed in #6173. 

An example of what would trigger this branch:

Initial State

Server:
```
foo
```

Client:
```
foo
```

request completions and type `b`

```
foob
```

Client:
```
foob
```

server sends back the following edit: `foo` -> `foobar` and `foo` -> `foober` (notice that the `b` will be removed by the savepoint) as a completion items, You select the first completion (`foobar`)  with tab but don't confirm the selection.

Server (purposefully out of sync):

```
foob
```

Client:
```
foobar
```

now wait for the completion re-request on idle timeout. This does not close the completion menu (that would be weird). Without this code we would request completions at the end of `foobar` which will be somewhere else in the document for the server which could cause all of the following:
* crash the client (out of range edit)
* cause an invalid edit to be applied
* just show no completions (so randomly removing `foober` from the list which is especially annoying with a low idle timeout and slow scrolling)
* crash the server immidietly (if the new cursor position is out of bounds)
* cause a permanent server/client desync making the LS misbehave until restarted

This is definitely an edgecase but it's important to correctly to account for to avoid really weird bugs

Reply:
Thanks for the thorough explanation!

> now wait for the completion re-request on idle timeout

I have not yet encountered this (was a thought I also had while reviewing)

AFAIK completion requests aren't sent on idle timeout when the completion menu is open (probably worth a PR though, had this feature in mind as well already).
Have you triggered this branch while testing? I currently can't think of a case where this branch is triggered.

Still I think it's safer to keep this, shouldn't hurt at least, when for whatever reason that edge case occurs where the completion menu is still open and a new completion request is sent, probably because of a new feature (like rerequesting completion items on idle-timeout when the completion menu is still open)

Reply:
I always assumed that was the case but that would explain some problems I have been running into... we really should be requesting completions especially to handle incomplete completion requests. I want to work on that soon anyway so at worst its a bit of future proofing